### PR TITLE
Allow user to retry OAuth2/OIDC login

### PR DESCRIPTION
**Pull request by fmoc**
screenshot_2022-08-16_10-48-33
screenshot_2022-08-16_10-49-05
screenshot_2022-08-16_10-49-12

I'm sure this could and should be polished. We are going to work on these widgets soon again when integrating them into the wizard, which is a good chance to improve the UX.

### CODE REVIEW

**Comment by TheOneRing** [comment]
Maybe we should move the icon to the top left?
Especially when the error is displayed it looks like its not aligned to anything?

**Comment by fmoc** [comment]
WIP in an upcoming PR. This just adds the retry functionality.

File: src/gui/creds/httpcredentialsgui.h [hunk 1]
```diff
@@ -64,7 +68,7 @@ private slots:
     void authorisationLinkChanged();
 
 private:
-    QScopedPointer<OAuth, QScopedPointerObjectDeleteLater<OAuth>> _asyncAuth;
+    QScopedPointer<AccountBasedOAuth, QScopedPointerObjectDeleteLater<OAuth>> _asyncAuth;
```
Comment (TheOneRing) [inline src/gui/creds/httpcredentialsgui.h:71]: Wrong class in the deleter

**Review by TheOneRing — CHANGES_REQUESTED** [review]
then merge

File: src/gui/loginrequireddialog/oauthloginrequiredwidget.cpp [hunk 2]
```diff
@@ -67,4 +85,25 @@ QList<QPair<QAbstractButton *, QDialogButtonBox::ButtonRole>> OAuthLoginRequired
     };
 }
 
+void OAuthLoginRequiredWidget::showErrorLabelAndRetryButton()
+{
+    _ui->errorFrame->show();
+
+    // apparently, we need to run this in the main event loop since the buttons are owned by the layout within the login required dialog
+    QTimer::singleShot(0, [this]() {
```
Comment (TheOneRing) [inline src/gui/loginrequireddialog/oauthloginrequiredwidget.cpp:93]: @erikjv  can you have a look with @fmoc why this hack is needed?
I guess there is a deeper issue here.

**Review by TheOneRing — CHANGES_REQUESTED** [review]

**Comment by TheOneRing** [comment]
Sadly the change is so outdated that I can't take over the development...

File: src/gui/loginrequireddialog/oauthloginrequiredwidget.cpp (hunk 2 above)
Comment (erikjv) [reply src/gui/loginrequireddialog/oauthloginrequiredwidget.cpp:93]: What is the order in which this method is called, and when `buttons()` is called? I mean, in the `buttons()` method, the ownership of the widget is transferred to the caller. So I'm wondering if this method is called before the layout adds the buttons to itself, and somehow invalidate its state.

**Comment by fmoc** [comment]
Need to rebase this PR on https://github.com/owncloud/client/pull/10176, which fixes a couple of problems.

File: src/gui/creds/httpcredentialsgui.cpp [hunk 3]
```diff
@@ -161,4 +152,18 @@ QUrl HttpCredentialsGui::authorisationLink() const
...
 
+void HttpCredentialsGui::restartOAuth()
+{
+    _asyncAuth.reset(new AccountBasedOAuth(_account->sharedFromThis(), this));
+    connect(_asyncAuth.data(), &OAuth::result,
+        this, &HttpCredentialsGui::asyncAuthResult);
+    connect(_asyncAuth.data(), &OAuth::destroyed,
+        this, &HttpCredentialsGui::authorisationLinkChanged);
+    connect(_asyncAuth.data(), &AccountBasedOAuth::refreshError, this, [this]() {
+        // FIXME
```
Comment (TheOneRing) [inline src/gui/creds/httpcredentialsgui.cpp:163]: ?

**Comment by fmoc** [comment]
screenshot_2022-10-11_17-31-30

Had to start over, but I managed to bring it into a usable state. Still investigating why the error label uses this much vertical space (Qt won't let me shrink it vertically during runtime at this point).

**Review by TheOneRing — APPROVED** [review]

**Pull request merged**